base/libgit2/callbacks.jl

```diff
@@ -65,82 +83,97 @@ function authenticate_ssh(libgit2credptr::Ptr{Ptr{Void}}, p::CredentialPayload,
+            else
+                creds.pubkey
+            end
         end
-    end
 
-    passphrase = Base.get(ENV, "SSH_KEY_PASS", creds.pass)
+        creds.pass = Base.get(ENV, "SSH_KEY_PASS", creds.pass)
 
-    if p.allow_prompt
+        p.use_env = false
+        modified = (
+            haskey(ENV, "SSH_KEY_PATH") ||
+            haskey(ENV, "SSH_PUB_KEY_PATH") ||
+            haskey(ENV, "SSH_KEY_PASS")
+        )
+    end
+
+    if p.remaining_prompts > 0 && (!modified || !isfilled(creds))
         # if username is not provided or empty, then prompt for it
+        username = username_ptr != Cstring(C_NULL) ? unsafe_string(username_ptr) : ""
         if isempty(username)
             prompt_url = git_url(scheme=p.scheme, host=p.host)
             response = Base.prompt("Username for '$prompt_url'", default=creds.user)
             isnull(response) && return user_abort()
-            username = unsafe_get(response)
+            creds.user = unsafe_get(response)
+        else
+            creds.user = username
         end
 
-        prompt_url = git_url(scheme=p.scheme, host=p.host, username=username)
+        prompt_url = git_url(scheme=p.scheme, host=p.host, username=creds.user)
 
         # For SSH we need a private key location
-        if !isfile(privatekey)
+        last_private_key = creds.prvkey
+        if !isfile(creds.prvkey) || !modified
             response = Base.prompt("Private key location for '$prompt_url'",
-                default=privatekey)
+                default=creds.prvkey)
             isnull(response) && return user_abort()
-            privatekey = unsafe_get(response)
+            creds.prvkey = expanduser(unsafe_get(response))
 
             # Only update the public key if the private key changed
-            if privatekey != creds.prvkey
-                publickey = privatekey * ".pub"
+            if creds.prvkey != last_private_key
+                creds.pubkey = creds.prvkey * ".pub"
             end
         end
 
         # For SSH we need a public key location. Avoid asking about the public key as
         # typically this will just annoy users.
-        if !isfile(publickey) && isfile(privatekey)
+        stale = !p.first_pass && creds.prvkey == last_private_key && creds.pubkey != creds.prvkey * ".pub"
+        if isfile(creds.prvkey) && (stale || !isfile(creds.pubkey))
             response = Base.prompt("Public key location for '$prompt_url'",
-                default=publickey)
+                default=creds.pubkey)
             isnull(response) && return user_abort()
-            publickey = unsafe_get(response)
+            creds.pubkey = expanduser(unsafe_get(response))
         end
 
-        if isempty(passphrase) && is_passphrase_required(privatekey)
+        # Ask for a passphrase when the private key exists and requires a passphrase
+        if isempty(creds.pass) && is_passphrase_required(creds.prvkey)
             if Sys.iswindows()
                 response = Base.winprompt(
```

Comment:
Move prompt logic into `ssh_auth_prompt`?

Reply:
Separating these into new functions hasn't turned out well. I've decided to stick with the status quo.